File: gui_apps/Doodle/step06/Line.cc
// Line.cc: implements custom line class
#include <QtGui>
#include <QList>
#include "Line.h"

Line::Line(int penWidth /*= 2*/, const QColor& penColor /*= qRgb(0,0,255)*/)
{
  _penWidth = penWidth;
  if (_penWidth <= 0) _penWidth = 2;
  _penColor = penColor;
  _points = new QList<QPoint>();
}

Line::~Line()
{
  delete _points;
}

void Line::setPenWidth(int newWidth)
{
  if (newWidth == _penWidth)
    return;
  _penWidth = (newWidth <= 0) ? 2 : newWidth;
  qDebug() << "Line::setPenWidth() -> new pen width: " << _penWidth;
  emit penWidthChanged(_penWidth);
}

void Line::setPenColor(const QColor& newColor)
{
  if (_penColor == newColor)
    return;
  _penColor = newColor;
  emit penColorChanged(_penColor);
}

int Line::numPoints() const
{
  return (_points == 0) ? 0 : _points->count();
}

void Line::addPoint(const QPoint& pt)
{
  if (_points == 0)
    _points = new QList<QPoint>();
  _points->append(pt);
}

void Line::draw(QPainter& painter)
{
  qDebug() << "In Line::draw()...";

  if (numPoints() > 0) {
    QPen pen(_penColor, _penWidth);
    painter.setPen(pen);
    
    bool first = true;
    QPoint lastPt;

    foreach(QPoint pt, *_points) {
      if (!first)
        painter.drawLine(lastPt, pt);
      else
        first = false;
      lastPt = pt;
    }
  }
}

File: gui_apps/PyDoodle/step05/line.py
"""
// ============================================================================
// line.py: represents one line in the Doodle, with its own color & width
//
// Tutorial - Qt Scribble Application
// Based on a similar tutorial for Borland ObjectWindows Library (OWL)
// Created by Manish Bhobe.
// My experiments with the Qt Framework. Use at your own risk!!
// ============================================================================
"""
import sys
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *

class Line(QObject):
    def __init__(self, penWidth=3, penColor=QColor(qRgb(0,65,255))):
        super(QObject, self).__init__()
        self.penWidth = min(max(2, penWidth), 12)
        self.penColor = penColor
        self.points = []

    def setPenWidth(self, penWidth):
        newWidth = min(max(2, penWidth), 12)
        if self.penWidth != newWidth:
            self.penWidth = newWidth

    def setPenColor(self, penColor):
        self.penColor = penColor

    def numPoints(self):
        return len(self.points)

    def append(self, pt: QPoint) -> None:
        self.points.append(pt)

    def draw(self, painter: QPainter) -> None:
        if self.numPoints() > 0:
            pen = QPen(self.penColor, self.penWidth)
            painter.setPen(pen)
            lastPt = None
            for i, pt in enumerate(self.points):
                if i > 0:
                    painter.drawLine(lastPt, pt)
                lastPt = pt
